nomad/state/state_store_secure_variables.go

```diff
@@ -0,0 +1,217 @@
+// registrations belonging to the provided namespace.
+func (s *StateStore) GetSecureVariablesByNamespace(
+	ws memdb.WatchSet, namespace string) (memdb.ResultIterator, error) {
+	txn := s.db.ReadTxn()
+
+	// Walk the entire table.
+	iter, err := txn.Get(TableSecureVariables, indexID+"_prefix", namespace, "")
+	if err != nil {
+		return nil, fmt.Errorf("secure variable lookup failed: %v", err)
+	}
+	ws.Add(iter.WatchCh())
+
+	return iter, nil
+}
+
+// GetSecureVariablesByNamespace returns an iterator that contains all
+// registrations belonging to the provided namespace.
+func (s *StateStore) GetSecureVariablesByNamespaceAndPrefix(
+	ws memdb.WatchSet, namespace, prefix string) (memdb.ResultIterator, error) {
+	txn := s.db.ReadTxn()
+
+	// Walk the entire table.
+	iter, err := txn.Get(TableSecureVariables, indexID+"_prefix", namespace, prefix)
+	if err != nil {
+		return nil, fmt.Errorf("secure variable lookup failed: %v", err)
+	}
+	ws.Add(iter.WatchCh())
+
+	return iter, nil
+}
+
+// GetSecureVariable returns an single secure variable at a given namespace and
+// path.
+func (s *StateStore) GetSecureVariable(
+	ws memdb.WatchSet, namespace, path string) (*structs.SecureVariable, error) {
+	txn := s.db.ReadTxn()
+
+	// Walk the entire table.
+	iter, err := txn.Get(TableSecureVariables, indexID, namespace, path)
+	if err != nil {
+		return nil, fmt.Errorf("secure variable lookup failed: %v", err)
+	}
+	ws.Add(iter.WatchCh())
+
+	for {
+		raw := iter.Next()
+		if raw == nil {
+			break
+		}
+		sv := raw.(*structs.SecureVariable)
+		return sv, nil
+	}
+	return nil, nil
+}
+
+func (s *StateStore) UpsertSecureVariables(msgType structs.MessageType, index uint64, svs []*structs.SecureVariable) error {
+	txn := s.db.WriteTxn(index)
+	defer txn.Abort()
+
+	var updated bool = false
+	for _, sv := range svs {
+		if err := s.upsertSecureVariableImpl(index, txn, sv, &updated); err != nil {
+			return err
+		}
+	}
+
+	if !updated {
+		return nil
+	}
+
+	if err := txn.Insert(tableIndex, &IndexEntry{TableSecureVariables, index}); err != nil {
+		return fmt.Errorf("index update failed: %v", err)
+	}
+
+	return txn.Commit()
+}
+
+// upsertSecureVariableImpl is used to upsert a namespace
+func (s *StateStore) upsertSecureVariableImpl(index uint64, txn *txn, svar *structs.SecureVariable, updated *bool) error {
+	sv := svar.Copy()
+
+	// TODO: Ensure the EncryptedData hash is non-nil. This should be done outside the state store
+	// for performance reasons, but we check here for defense in depth.
+	// if len(sv.Hash) == 0 {
+	// 	sv.SetHash()
+	// }
+
+	// For maximum safety, nil UnencryptedData
+	sv.UnencryptedData = nil
+
+	// Check if the secure variable already exists
+	existing, err := txn.First(TableSecureVariables, indexID, sv.Namespace, sv.Path)
+	if err != nil {
+		return fmt.Errorf("secure variable lookup failed: %v", err)
+	}
+
+	sv.Canonicalize()
+
+	// Setup the indexes correctly
+	now := time.Now().Round(0)
```

Comment:
To eliminate the realtime clock information. I was having challenges with equality checking otherwise.  It could have been a red herring though.  Let me do some testing without it.